frontend/src/metabase-lib/v1/expressions/tokenizer/tokens.ts

```diff
@@ -0,0 +1,95 @@
+import { ExternalTokenizer } from "@lezer/lr";
+
+import { Reference } from "./parser.terms";
+
+function char(char: string): number {
+  if (char.length !== 1) {
+    throw new Error(`Expected a single character, got ${char}`);
+  }
+  return char.charCodeAt(0);
+}
+
+const BACKSLASH = char("\\");
+const OPEN_BRACKET = char("[");
+const CLOSE_BRACKET = char("]");
+const NEW_LINE = char("\n");
+const EOF = -1;
+
+const OPERATORS = new Set([
+  OPEN_BRACKET,
+  char(","),
+  char(" "),
+  char("("),
+  char(")"),
+]);
+
+function isOperator(char: number) {
+  return OPERATORS.has(char);
+}
+
+/**
+ * Reference (or bracket idenfiers) like `[User Id]` are parsed differently
+ * based on whether they are well-formed or not.
+ *
+ * This is hard to express in the grammar, so we use this tokenizer to match them.
+ */
+export const reference = new ExternalTokenizer(input => {
```

Comment:
This is a bit unfortunate, but the current grammar for bracket identifiers is a bit weird and hard to encode in the grammar:
- when a bracket identifier is closed, it can contain any content, ie. `[Foo, _ + - \[\]()]`
- when a bracket identifier is not closed properly (ie. missing a closing `]`), it ends at the first operator, ie. `[Foo, bar` only should match `[Foo`

This was done to make autocompletion not eat the whole line and AFAICT does not have other uses.